File: backend/pkg/api/page/handlers.go
package page

import (
	"sync"

	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	corePage "backend/pkg/core/page"
	"backend/pkg/database"
	"backend/pkg/database/repository"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

var (
	serviceInstance *corePage.Service
	serviceOnce     sync.Once
)

// getService returns the page service instance (lazily initialized)
func getService() *corePage.Service {
	serviceOnce.Do(func() {
		db := database.DB()
		pageRepo := repository.NewPageRepository(db)
		serviceInstance = corePage.NewService(pageRepo)
	})
	return serviceInstance
}

// GetPageBySlug handles GET /pages/:slug
// @Summary Get page by slug
// @Description Get a public page by its slug
// @Tags pages
// @Accept json
// @Produce json
// @Param slug path string true "Page slug"
// @Success 200 {object} response.SuccessResponse{data=dto.PageResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /pages/{slug} [get]
func GetPageBySlug(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Page slug is required"))
	}

	svc := getService()
	page, err := svc.GetBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, page)
}

// ListPages handles GET /dashboard/pages
// @Summary List pages
// @Description Get a paginated list of pages (dashboard)
// @Tags pages
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param perPage query int false "Items per page" default(20)
// @Param isPublished query bool false "Filter by published status"
// @Success 200 {object} response.SuccessResponse{data=dto.PageListResponse}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/pages [get]
func ListPages(c *fiber.Ctx) error {
	pageNum := c.QueryInt("page", 1)
	perPage := c.QueryInt("perPage", 20)

	var isPublished *bool
	isPublishedStr := c.Query("isPublished")
	if isPublishedStr != "" {
		val := c.QueryBool("isPublished", true)
		isPublished = &val
	}

	svc := getService()
	result, err := svc.List(c.Context(), pageNum, perPage, isPublished)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, result)
}

// GetPageByID handles GET /dashboard/pages/:id
// @Summary Get page by ID
// @Description Get a page by its ID (dashboard)
// @Tags pages
// @Accept json
// @Produce json
// @Param id path string true "Page ID"
// @Success 200 {object} response.SuccessResponse{data=dto.PageResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/pages/{id} [get]
func GetPageByID(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid page ID"))
	}

	svc := getService()
	page, err := svc.GetByID(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, page)
}

// CreatePage handles POST /dashboard/pages
// @Summary Create page
// @Description Create a new page
// @Tags pages
// @Accept json
// @Produce json
// @Param request body dto.CreatePageRequest true "Page details"
// @Success 201 {object} response.SuccessResponse{data=dto.PageResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/pages [post]
func CreatePage(c *fiber.Ctx) error {
	var req corePage.CreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	svc := getService()
	page, err := svc.Create(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, page)
}

// UpdatePage handles PUT /dashboard/pages/:id
// @Summary Update page
// @Description Update an existing page
// @Tags pages
// @Accept json
// @Produce json
// @Param id path string true "Page ID"
// @Param request body dto.UpdatePageRequest true "Page update details"
// @Success 200 {object} response.SuccessResponse{data=dto.PageResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/pages/{id} [put]
func UpdatePage(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid page ID"))
	}

	var req corePage.UpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	svc := getService()
	page, err := svc.Update(c.Context(), id, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, page)
}

// DeletePage handles DELETE /dashboard/pages/:id
// @Summary Delete page
// @Description Delete a page by ID
// @Tags pages
// @Accept json
// @Produce json
// @Param id path string true "Page ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /dashboard/pages/{id} [delete]
func DeletePage(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid page ID"))
	}

	svc := getService()
	if err := svc.Delete(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

File: backend/src/api/page/handlers.rs
use axum::{
    Json,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::api::response::SuccessResponse;

use super::{
    dto::{
        PageCreateRequest, PageListQuery, PageListResponse, PageResponse, PageUpdateRequest,
        SuccessFlag,
    },
    error::{PageApiError, PageAuthenticated},
    state::PageState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, PageApiError>;

fn page_id(value: &str) -> Result<Uuid, PageApiError> {
    Uuid::parse_str(value)
        .map_err(|_| crate::error::AppError::InvalidInput("Invalid page ID".to_owned()).into())
}

#[utoipa::path(
    get,
    path = "/pages/{slug}",
    params(("slug" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<PageResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    tag = "pages",
    operation_id = "getPageBySlug"
)]
pub async fn get_page_by_slug(
    State(state): State<PageState>,
    Path(slug): Path<String>,
) -> ApiResult<PageResponse> {
    if slug.is_empty() {
        return Err(
            crate::error::AppError::InvalidInput("Page slug is required".to_owned()).into(),
        );
    }
    Ok(Json(SuccessResponse::new(
        state.service().get_by_slug(&slug).await?.into(),
    )))
}

#[utoipa::path(
    get,
    path = "/dashboard/pages",
    params(PageListQuery),
    responses(
        (status = 200, body = SuccessResponse<PageListResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "pages",
    operation_id = "listPages"
)]
pub async fn list_pages(
    _authenticated: PageAuthenticated,
    State(state): State<PageState>,
    Query(query): Query<PageListQuery>,
) -> ApiResult<PageListResponse> {
    let (page, per_page, is_published) = query.values();
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .list(page, per_page, is_published)
            .await?
            .into(),
    )))
}

#[utoipa::path(
    get,
    path = "/dashboard/pages/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<PageResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "pages",
    operation_id = "getPageById"
)]
pub async fn get_page_by_id(
    _authenticated: PageAuthenticated,
    State(state): State<PageState>,
    Path(id): Path<String>,
) -> ApiResult<PageResponse> {
    Ok(Json(SuccessResponse::new(
        state.service().get_by_id(page_id(&id)?).await?.into(),
    )))
}

#[utoipa::path(
    post,
    path = "/dashboard/pages",
    request_body = PageCreateRequest,
    responses(
        (status = 201, body = SuccessResponse<PageResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 409, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "pages",
    operation_id = "createPage"
)]
pub async fn create_page(
    _authenticated: PageAuthenticated,
    State(state): State<PageState>,
    body: Result<Json<PageCreateRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<PageResponse>>), PageApiError> {
    let Json(request) = body.map_err(|_| PageApiError::invalid_body())?;
    let page = state.service().create(request.validate()?).await?;
    Ok((StatusCode::CREATED, Json(SuccessResponse::new(page.into()))))
}

#[utoipa::path(
    put,
    path = "/dashboard/pages/{id}",
    params(("id" = Uuid, Path)),
    request_body = PageUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<PageResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "pages",
    operation_id = "updatePage"
)]
pub async fn update_page(
    _authenticated: PageAuthenticated,
    State(state): State<PageState>,
    Path(id): Path<String>,
    body: Result<Json<PageUpdateRequest>, JsonRejection>,
) -> ApiResult<PageResponse> {
    let id = page_id(&id)?;
    let Json(request) = body.map_err(|_| PageApiError::invalid_body())?;
    Ok(Json(SuccessResponse::new(
        state.service().update(id, request.into()).await?.into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/dashboard/pages/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "pages",
    operation_id = "deletePage"
)]
pub async fn delete_page(
    _authenticated: PageAuthenticated,
    State(state): State<PageState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    state.service().delete(page_id(&id)?).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}
